### PR TITLE
[Feature] Result🏆  Entity Listener👂

### PR DESCRIPTION
**Pull request by stefanwaldhauser**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest2.ase.in.tum.de/ **(I have db access here and can check that the table is correctly updated)**
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Server: I added multiple integration tests (Spring) related to the features (with a high test coverage)
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.

### Motivation and Context

Problem: Getting the last result of a student for a given exercise requires a multi step query (first we have to get the participation, then the last submission, then the last result of that submission). Therefore various calculations based on these last results take a long time. 🐌🐌🐌🐌🐌🐌🐌

### Description

Solution to Problem: 
- Create new table called 'participant_score' where we save the last result and last rated result (meaning it was before the exercise deadline) for a given exercise and student / team.
- This table gets pre-filled with a migration script
- When Artemis runs a so called "EntityListener" (https://docs.jboss.org/hibernate/stable/entitymanager/reference/en/html/listeners.html) is hooked into the Result lifecycle and whenever a new result comes in or an old one is updated, the table gets also updated.
- Querys on this table should be much faster 🦔 🦔 🦔 🦔 🦔 🦔 🦔 

Scenario to make things more clearer:
Imagine an exercise with the id 1 and a student with the id 1. The student now solves the exercise before the due date with a perfect score and thus a new **rated result** with id 1 is generated. The entity listener now watches for this creation event and inserts the following row into the table:
image
Imagine that the same student now tries the exercise as exam preparation again  but AFTER the exercise due date. he achieves a score of 50 points thus a new **unrated result** with id 2 is generated. The entity listener now watches for this creation event and updates the following row in the table:
image
Now imagine that the instructor manually updates the score of the second result to 70. . The entity listener now watches for this update event and updates the following row in the table:
image

Structure of the table:
image

### Steps for Testing

**Testers WITH a local setup where they can see the database**
Play around with Artemis and different scenarios (creating result, re-calculate quiz, recalculate programming exercise and so on) and check that the student score is consistent and gets updated to the correct result!

Also check that the migration script filled in the participant scores table correctly.

**Testers WITHOUT a local setup where they can see the database**
Play around with Artemis and different scenarios (creating result, re-calculate quiz, recalculate programming exercise and so on) and check the participant score table in the client (see screenshots)

**Please test going through an exam especially well with all exercise types. nothing should be broken here**

### Test Coverage

image
image

### Screenshots

**How to get to the participant scores page of a course**
image
image
image

**How to get to the participant scores page of an exam**
image
image
image

### CODE REVIEW

File: src/test/resources/config/application-artemis.yml [hunk 1]
```diff
@@ -59,3 +59,5 @@ artemis:
         feedback-consistency-url: http://localhost:8001/feedback_consistency
         base64-secret: YWVuaXF1YWRpNWNlaXJpNmFlbTZkb283dXphaVF1b29oM3J1MWNoYWlyNHRoZWUzb2huZ2FpM211bGVlM0VpcAo=
         token-validity-in-seconds: 10800
+    resultListener:
```
Comment (stefanwaldhauser) [inline src/test/resources/config/application-artemis.yml:62]: Why is this needed?
- The problem was that during tests a lot of tests failed as **after making a REST call the mock mvc removes the authentication object meaning no changing sql statements (create, update, delete) are possible any more**. Thus the entity listener needs to be deactivated during tests.

File: src/main/resources/config/application-artemis.yml [hunk 2]
```diff
@@ -88,3 +88,5 @@ artemis:
         feedback-consistency-url: http://localhost:8001/feedback_consistency
         base64-secret: YWVuaXF1YWRpNWNlaXJpNmFlbTZkb283dXphaVF1b29oM3J1MWNoYWlyNHRoZWUzb2huZ2FpM211bGVlM0VpcAo=
         token-validity-in-seconds: 10800
+    resultListener:
```
Comment (stefanwaldhauser) [inline src/main/resources/config/application-artemis.yml:91]: Why is this needed?

The problem was that during tests a lot of tests failed as after making a REST call the mock mvc removes the authentication object meaning no changing sql statements (create, update, delete) are possible any more. Thus the entity listener needs to be deactivated during tests.

File: src/main/java/de/tum/in/www1/artemis/service/listeners/ResultListener.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+    /**
+     * Update or create a new participation score after a result is created or updated
+     *
+     * @param result created or updated result
+     */
+    @PostUpdate
+    @PostPersist
+    public void updateOrCreateParticipantScore(Result result) {
+        if (isActive == null || !isActive) {
```
Comment (stefanwaldhauser) [inline src/main/java/de/tum/in/www1/artemis/service/listeners/ResultListener.java:55]: Why is this needed?

The problem was that during tests a lot of tests failed as after making a REST call the mock mvc removes the authentication object meaning no changing sql statements (create, update, delete) are possible any more. Thus the entity listener needs to be deactivated during tests.

File: src/main/java/de/tum/in/www1/artemis/service/listeners/ResultListener.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    /**
+     * Remove or update associated participation scores before a result is removed
+     *
+     * @param resultToBeDeleted result about to be remove
+     */
+    @PreRemove
+    public void removeOrUpdateAssociatedParticipantScore(Result resultToBeDeleted) {
+        if (isActive == null || !isActive) {
```
Comment (stefanwaldhauser) [inline src/main/java/de/tum/in/www1/artemis/service/listeners/ResultListener.java:40]: Why is this needed?

The problem was that during tests a lot of tests failed as after making a REST call the mock mvc removes the authentication object meaning no changing sql statements (create, update, delete) are possible any more. Thus the entity listener needs to be deactivated during tests.

File: src/main/java/de/tum/in/www1/artemis/service/listeners/ResultListener.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+import org.springframework.context.annotation.Lazy;
+import org.springframework.stereotype.Component;
+
+import de.tum.in.www1.artemis.domain.Result;
+import de.tum.in.www1.artemis.service.ScoreService;
+
+@Component
+public class ResultListener {
+
+    @Value("#{new Boolean('${artemis.resultListener.isActive}')}")
```
Comment (stefanwaldhauser) [inline src/main/java/de/tum/in/www1/artemis/service/listeners/ResultListener.java:17]: Why is this needed?

The problem was that during tests a lot of tests failed as after making a REST call the mock mvc removes the authentication object meaning no changing sql statements (create, update, delete) are possible any more. Thus the entity listener needs to be deactivated during tests.

File: src/test/java/de/tum/in/www1/artemis/util/ProgrammingExerciseTestService.java [hunk 6]
```diff
@@ -334,7 +333,6 @@ public void importExercise_created(ProgrammingLanguage programmingLanguage, bool
         // Import the exercise and load all referenced entities
         var importedExercise = request.postWithResponseBody(ROOT + IMPORT.replace("{sourceExerciseId}", sourceExercise.getId().toString()), exerciseToBeImported,
                 ProgrammingExercise.class, params, HttpStatus.OK);
-        SecurityUtils.setAuthorizationObject();
```
Comment (stefanwaldhauser) [inline src/test/java/de/tum/in/www1/artemis/util/ProgrammingExerciseTestService.java:337]: All these are not needed anymore as I modified @RequestUtil

File: src/main/java/de/tum/in/www1/artemis/service/ScoreService.java [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+     * @param createdOrUpdatedResult newly created or updated result
+     */
+    @Transactional(propagation = Propagation.NOT_SUPPORTED)
+    public void updateOrCreateParticipantScore(Result createdOrUpdatedResult) {
+        if (createdOrUpdatedResult.getScore() == null || createdOrUpdatedResult.getCompletionDate() == null || createdOrUpdatedResult.getParticipation() == null) {
+            return;
+        }
+
+        // There is a deadlock problem with programming exercises here if we use the participation from the result (reason unknown at the moment)
+        // ,therefore we get the participation from the database
```
Comment (derLalla) [inline src/main/java/de/tum/in/www1/artemis/service/ScoreService.java:124]:
```suggestion
        // therefore we get the participation from the database
```

File: src/main/java/de/tum/in/www1/artemis/service/ScoreService.java [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+        Participation participation = participationOptional.get();
+
+        boolean isStudentParticipation = participation.getClass().equals(StudentParticipation.class)
+                || participation.getClass().equals(ProgrammingExerciseStudentParticipation.class);
+        if (!isStudentParticipation) {
+            return Optional.empty();
+        }
+        else {
+            return Optional.of((StudentParticipation) participation);
+        }
```
Comment (derLalla) [inline src/main/java/de/tum/in/www1/artemis/service/ScoreService.java:161]:
```suggestion
        return Optional.of((StudentParticipation) participation);
```

File: src/main/java/de/tum/in/www1/artemis/service/ScoreService.java [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+
+    private void setLastResultAttributes(ParticipantScore associatedParticipantScore, Result newLastResult) {
+        associatedParticipantScore.setLastResult(newLastResult);
+        if (newLastResult == null) {
+            associatedParticipantScore.setLastScore(null);
+        }
+        else {
+            associatedParticipantScore.setLastScore(newLastResult.getScore());
+        }
+
```
Comment (derLalla) [inline src/main/java/de/tum/in/www1/artemis/service/ScoreService.java:309]:
```suggestion
```

File: src/main/java/de/tum/in/www1/artemis/service/ScoreService.java [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+    private void setLastRatedResultAttributes(ParticipantScore associatedParticipantScore, Result newLastRatedResult) {
+        associatedParticipantScore.setLastRatedResult(newLastRatedResult);
+        if (newLastRatedResult == null) {
+            associatedParticipantScore.setLastRatedScore(null);
+        }
+        else {
+            associatedParticipantScore.setLastRatedScore(newLastRatedResult.getScore());
+        }
+    }
+
```
Comment (derLalla) [inline src/main/java/de/tum/in/www1/artemis/service/ScoreService.java:321]:
```suggestion
```

File: src/main/java/de/tum/in/www1/artemis/service/listeners/ResultListener.java [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+     * <p>
+     * Will be called by Hibernate AFTER a result is updated or created
+     *
+     * @param createdOrUpdatedResult created or updated result
+     */
+    @PostUpdate
+    @PostPersist
+    public void updateOrCreateParticipantScore(Result createdOrUpdatedResult) {
+        scoreService.updateOrCreateParticipantScore(createdOrUpdatedResult);
+
```
Comment (derLalla) [inline src/main/java/de/tum/in/www1/artemis/service/listeners/ResultListener.java:60]:
```suggestion
```

File: src/test/java/de/tum/in/www1/artemis/entitylistener/ResultListenerIntegrationTest.java [hunk 12]
```diff
@@ -0,0 +1,506 @@
...
+        // creating the users student1-student5, tutor1-tutor10 and instructors1-instructor10
+        this.database.addUsers(5, 10, 10);
+        User student1 = userRepository.findOneByLogin("student1").get();
+        idOfStudent1 = student1.getId();
+        // creating course
+        Course course = this.database.createCourse();
+        idOfCourse = course.getId();
+        createIndividualTextExercise(pastTimestamp, pastTimestamp, pastTimestamp);
+        createTeamTextExerciseAndTeam(pastTimestamp, pastTimestamp, pastTimestamp);
+
```
Comment (derLalla) [inline src/test/java/de/tum/in/www1/artemis/entitylistener/ResultListenerIntegrationTest.java:99]:
```suggestion
```

File: src/test/java/de/tum/in/www1/artemis/entitylistener/ResultListenerIntegrationTest.java [hunk 13]
```diff
@@ -0,0 +1,506 @@
...
+        ParticipantScore originalParticipantScore = setupTestScenarioWithOneResultSaved(false, isTeamTest);
+        Result originalResult = originalParticipantScore.getLastResult();
+        // creating a new rated result should trigger the entity listener and update the student score BUT only the not rated part
+        Result newResult = createNewResult(isTeamTest, false);
+        verifyStructureOfParticipantScoreInDatabase(isTeamTest, newResult.getId(), newResult.getScore(), null, null);
+        resultRepository.deleteById(newResult.getId());
+        List<Result> savedResults = resultRepository.findAll();
+        assertThat(savedResults).size().isEqualTo(1);
+        verifyStructureOfParticipantScoreInDatabase(isTeamTest, originalResult.getId(), originalResult.getScore(), null, null);
+
```
Comment (derLalla) [inline src/test/java/de/tum/in/www1/artemis/entitylistener/ResultListenerIntegrationTest.java:170]:
```suggestion
```

File: src/test/resources/config/application.yml [hunk 14]
```diff
@@ -21,6 +21,7 @@ artemis:
     file-upload-path: uploads
     submission-export-path: exports
 
+
```
Comment (derLalla) [inline src/test/resources/config/application.yml:24]:
```suggestion
```

**Review by derLalla — APPROVED** [review]
this was a nice PR to look through the code. Excellent job. I only have some little adjustments, which are not that critical. So I am approving.

File: src/main/java/de/tum/in/www1/artemis/web/rest/ParticipantScoreResource.java [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+     *
+     * @param courseId   the id of the course to which the learning goal belongs
+     * @param pageable   pageable object to filter and
+     * @param getUnpaged if set all participant scores of the course will be loaded
+     * @return the ResponseEntity with status 200 (OK) and with the learning goal course performance in the body
+     */
+    @GetMapping("/courses/{courseId}/participant-scores")
+    @PreAuthorize("hasAnyRole('INSTRUCTOR', 'ADMIN')")
+    public ResponseEntity<List<ParticipantScoreDTO>> getParticipantScoresOfCourse(@PathVariable Long courseId, Pageable pageable,
+            @RequestParam(value = "getUnpaged", required = false, defaultValue = "false") Boolean getUnpaged) {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ParticipantScoreResource.java:59]:
```suggestion
            @RequestParam(value = "getUnpaged", required = false, defaultValue = "false") boolean getUnpaged) {
```

**Review by derLalla — DISMISSED** [review]
### Old Exercise 
Setting the first result to 100%:
image

Then overwriting the assessment with 44%:
image

Result after the due date:
image
Result after complaint assessment:
image

### New Exercise
Result for empty submit:
image

Result after next submit with 8%:
image

Result for build failed:
image

Result for fixing build and solving next test case:
image
Empty submit:
image

After manual assessing:
image


### Quiz Exercise:
After participation:
image

Re-evaluating quiz and set a question to invalid (now 100% correct):
image

Set quiz to practice and participate in it (result id has correctly changed and last rated stayed the same):
image

### Total
The outcomes were all expected and worked as described! 👍

File: src/main/webapp/app/exam/manage/exam-participant-scores/exam-participant-scores.component.html [hunk 16]
```diff
@@ -0,0 +1,12 @@
...
+<div *ngIf="isLoading" class="d-flex justify-content-center">
+    <div class="spinner-border" role="status">
+        <span class="sr-only">{{ 'loading' | translate }}</span>
+    </div>
+</div>
+
+<div *ngIf="!isLoading && participantScores">
+    <button class="btn btn-info mb-2" (click)="loadData()">Refresh</button>
+    <jhi-participant-scores [isLoading]="isLoading" [participantScores]="participantScores"></jhi-participant-scores>
+</div>
```
Comment (iozbeyli) [inline src/main/webapp/app/exam/manage/exam-participant-scores/exam-participant-scores.component.html:12]: This is same with course-participant-scores.component.html so a common child component would be nice for this.

File: src/main/webapp/app/shared/participant-scores/participant-scores.component.ts [hunk 17]
```diff
@@ -0,0 +1,21 @@
+import { Component, Input } from '@angular/core';
+import { ParticipantScoreDTO } from 'app/shared/participant-scores/participant-scores.service';
+
+@Component({
+    selector: 'jhi-participant-scores',
+    templateUrl: './participant-scores.component.html',
+    styles: [],
```
Comment (iozbeyli) [inline src/main/webapp/app/shared/participant-scores/participant-scores.component.ts:7]: No style is applied, this can be removed.

File: src/main/webapp/app/shared/participant-scores/participant-scores.component.ts [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+export class ParticipantScoresComponent {
+    @Input()
+    participantScores: ParticipantScoreDTO[] = [];
+    @Input()
+    isLoading = false;
+    extractParticipantName = (participantScoreDTO: ParticipantScoreDTO) => {
+        if (participantScoreDTO.userName) {
+            return `${participantScoreDTO.userName}`;
+        } else {
+            return `${participantScoreDTO.teamName}`;
```
Comment (iozbeyli) [inline src/main/webapp/app/shared/participant-scores/participant-scores.component.ts:18]: What if there is no teamName?

File: src/main/webapp/app/shared/participant-scores/participant-scores.service.ts [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+        });
+    }
+
+    findAllOfExamPaged(examId: number, sortingParameters: SortingParameter[], page: number, size: number): Observable<HttpResponse<ParticipantScoreDTO[]>> {
+        let params = new HttpParams();
+        for (const sortParameter of sortingParameters) {
+            params = params.append('sort', `${sortParameter.sortProperty},${sortParameter.sortDirection}`);
+        }
+        params = params.set('page', page + '');
+        params = params.set('size', size + '');
```
Comment (iozbeyli) [inline src/main/webapp/app/shared/participant-scores/participant-scores.service.ts:65]: It is almost same with findAllOfCoursePaged so the common parts can be moved to a function.

File: src/main/webapp/app/exam/manage/exam-participant-scores/exam-participant-scores.component.html (hunk 16 above)
Comment (stefanwaldhauser) [reply src/main/webapp/app/exam/manage/exam-participant-scores/exam-participant-scores.component.html:18]: done

File: src/main/webapp/app/shared/participant-scores/participant-scores.component.ts (hunk 17 above)
Comment (stefanwaldhauser) [reply src/main/webapp/app/shared/participant-scores/participant-scores.component.ts]: done

File: src/main/webapp/app/shared/participant-scores/participant-scores.component.ts (hunk 18 above)
Comment (stefanwaldhauser) [reply src/main/webapp/app/shared/participant-scores/participant-scores.component.ts]: This can never happen. Either a result is generated from a student or from a team

File: src/main/webapp/app/shared/participant-scores/participant-scores.service.ts (hunk 19 above)
Comment (stefanwaldhauser) [reply src/main/webapp/app/shared/participant-scores/participant-scores.service.ts]: Done

**Review by tiffca — DISMISSED** [review]
Tested successfully, participating in exercises adds to the table. Participating in and assessing exams adds to the table

File: src/main/webapp/app/shared/participant-scores/participant-scores-average-table/participant-scores-average-table.component.ts [hunk 20]
```diff
@@ -0,0 +1,21 @@
...
+    participantAverageScores: ParticipantScoreAverageDTO[] = [];
+    @Input()
+    isLoading = false;
+
+    extractParticipantName = (participantScoreAverageDTO: ParticipantScoreAverageDTO) => {
+        if (participantScoreAverageDTO.userName) {
+            return `${participantScoreAverageDTO.userName}`;
+        } else {
+            return `${participantScoreAverageDTO.teamName}`;
+        }
```
Comment (derLalla) [inline src/main/webapp/app/shared/participant-scores/participant-scores-average-table/participant-scores-average-table.component.ts:19]: remove the else:

```suggestion
        return `${participantScoreAverageDTO.teamName}`;
```

When username is empty, is teamName always set? Otherwise we would need a default return value

File: src/main/webapp/app/shared/participant-scores/participant-scores.service.ts [hunk 21]
```diff
@@ -21,6 +21,17 @@ export class ParticipantScoreDTO {
...
 
+/**
+ * Normally the use of null in the client is discouraged but in this case it is fine as the server will always send all
+ * the properties but will give them the value null (a property will never be undefined here).
+ */
+export class ParticipantScoreAverageDTO {
+    public userName: string | null;
+    public teamName: string | null;
+    public averageScore: number | null;
+    public averageRatedScore: number | null;
```
Comment (derLalla) [inline src/main/webapp/app/shared/participant-scores/participant-scores.service.ts:32]: are you sure you need `null` instead of `undefined`?

File: src/main/webapp/app/shared/participant-scores/participant-scores.service.ts [hunk 22]
```diff
@@ -0,0 +1,120 @@
...
+    public userId: number | null;
+    public userName: string | null;
+    public teamId: number | null;
+    public teamName: string | null;
+    public exerciseId: number | null;
+    public exerciseTitle: string | null;
+    public lastResultId: number | null;
+    public lastResultScore: number | null;
+    public lastRatedResultId: number | null;
+    public lastRatedResultScore: number | null;
```
Comment (derLalla) [inline src/main/webapp/app/shared/participant-scores/participant-scores.service.ts:21]: are you sure you need to use `null` instead of `undefined`?

File: src/main/webapp/app/shared/participant-scores/participant-scores.service.ts [hunk 23]
```diff
@@ -0,0 +1,120 @@
...
+
+/**
+ * Normally the use of null in the client is discouraged but in this case it is fine as the server will always send all
+ * the properties but will give them the value null (a property will never be undefined here).
+ */
+export class ParticipantScoreAverageDTO {
+    public userName: string | null;
+    public teamName: string | null;
+    public averageScore: number | null;
+    public averageRatedScore: number | null;
```
Comment (derLalla) [inline src/main/webapp/app/shared/participant-scores/participant-scores.service.ts:32]: also here?

File: src/main/java/de/tum/in/www1/artemis/web/rest/dto/ParticipantScoreAverageDTO.java [hunk 24]
```diff
@@ -0,0 +1,30 @@
...
+
+    public ParticipantScoreAverageDTO(User user, Double averageScore, Double averageRatedScore) {
+        this.userName = user.getLogin();
+        this.teamName = null;
+        this.averageScore = averageScore;
+        this.averageRatedScore = averageRatedScore;
+    }
+
+    public ParticipantScoreAverageDTO(Team team, Double averageScore, Double averageRatedScore) {
+        this.userName = null;
```
Comment (derLalla) [inline src/main/java/de/tum/in/www1/artemis/web/rest/dto/ParticipantScoreAverageDTO.java:24]: you are using null instead of undefined. Is this intended?

Comment (derLalla) [inline src/main/java/de/tum/in/www1/artemis/web/rest/dto/ParticipantScoreAverageDTO.java:24]: you are using null instead of undefined. Is this intended?

File: src/main/webapp/app/shared/participant-scores/participant-scores.service.ts [hunk 25]
```diff
@@ -0,0 +1,120 @@
...
+        });
+    }
+
+    private createPageParameters(sortingParameters: SortingParameter[], page: number, size: number) {
+        let params = new HttpParams();
+        for (const sortParameter of sortingParameters) {
+            params = params.append('sort', `${sortParameter.sortProperty},${sortParameter.sortDirection}`);
+        }
+        params = params.set('page', page + '');
+        return params.set('size', size + '');
```
Comment (derLalla) [inline src/main/webapp/app/shared/participant-scores/participant-scores.service.ts:118]:
```suggestion
        params = params.set('page', String(page));
        return params.set('size', String(size));
```

**Review by derLalla — CHANGES_REQUESTED** [review]
I have noted some changes, besides that are the new changes very good.

File: src/main/webapp/app/shared/participant-scores/participant-scores-average-table/participant-scores-average-table.component.ts (hunk 20 above)
Comment (stefanwaldhauser) [reply src/main/webapp/app/shared/participant-scores/participant-scores-average-table/participant-scores-average-table.component.ts:19]: Yes either the team name or the userName are set

File: src/main/webapp/app/shared/participant-scores/participant-scores.service.ts (hunk 21 above)
Comment (stefanwaldhauser) [reply src/main/webapp/app/shared/participant-scores/participant-scores.service.ts:32]: Yes. In this case here yes. I also wrote the following comment:
`/**
 * Normally the use of null in the client is discouraged but in this case it is fine as the server will always send all
 * the properties but will give them the value null (a property will never be undefined here).
 */
export class ParticipantScoreDTO {
    public id: number | null;
    public userId: number | null;
    public userName: string | null;
    public teamId: number | null;
    public teamName: string | null;
    public exerciseId: number | null;
    public exerciseTitle: string | null;
    public lastResultId: number | null;
    public lastResultScore: number | null;
    public lastRatedResultId: number | null;
    public lastRatedResultScore: number | null;
}`

File: src/main/webapp/app/shared/participant-scores/participant-scores.service.ts (hunk 22 above)
Comment (stefanwaldhauser) [reply src/main/webapp/app/shared/participant-scores/participant-scores.service.ts:21]: Yes. In this case here yes. I wrote the server code such that all the properties are sent always. The object is used in the table as is, just dropping some properties makes it harder to just display everything in a table. (i would have to always handle the case that every single property could potentially be undefined and so on)

File: src/main/webapp/app/shared/participant-scores/participant-scores.service.ts (hunk 23 above)
Comment (stefanwaldhauser) [reply src/main/webapp/app/shared/participant-scores/participant-scores.service.ts:32]: Yes. In this case here yes. I wrote the server code such that all the properties are sent always. The object is used in the table as is, just dropping some properties makes it harder to just display everything in a table. (i would have to always handle the case that every single property could potentially be undefined and so on)

File: src/main/java/de/tum/in/www1/artemis/web/rest/dto/ParticipantScoreAverageDTO.java (hunk 24 above)
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/web/rest/dto/ParticipantScoreAverageDTO.java:24]: Yes. In this case here yes. I wrote the server code such that all the properties are sent always. The object is used in the table as is, just dropping some properties makes it harder to just display everything in a table. (i would have to always handle the case that every single property could potentially be undefined and so on)

Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/web/rest/dto/ParticipantScoreAverageDTO.java:24]: Yes. In this case here yes. I wrote the server code such that all the properties are sent always. The object is used in the table as is, just dropping some properties makes it harder to just display everything in a table. (i would have to always handle the case that every single property could potentially be undefined and so on)

**Review by derLalla — DISMISSED** [review]
re-approving code changes!

**Review by tiffca — COMMENTED** [review]
Tested successfully, scores show as expected in participant scores table for regular exercises and exams

**Review by tiffca — APPROVED** [review]

**Review by LudwigWeide — DISMISSED** [review]
Tested on TS1. Approve.

File: src/main/webapp/app/shared/participant-scores/participant-scores-average-table/participant-scores-average-table.component.ts [hunk 26]
```diff
@@ -0,0 +1,20 @@
...
+    @Input()
+    participantAverageScores: ParticipantScoreAverageDTO[] = [];
+    @Input()
+    isLoading = false;
+
+    extractParticipantName = (participantScoreAverageDTO: ParticipantScoreAverageDTO) => {
+        if (participantScoreAverageDTO.userName) {
+            return `${participantScoreAverageDTO.userName}`;
+        }
+        return `${participantScoreAverageDTO.teamName}`;
```
Comment (derLalla) [inline src/main/webapp/app/shared/participant-scores/participant-scores-average-table/participant-scores-average-table.component.ts:18]:
```suggestion
        return participantScoreAverageDTO.userName ? String(participantScoreAverageDTO.userName) : String(participantScoreAverageDTO.teamName);
```

File: src/main/webapp/app/shared/participant-scores/participant-scores-table/participant-scores-table.component.ts [hunk 27]
```diff
@@ -0,0 +1,21 @@
...
+    participantScores: ParticipantScoreDTO[] = [];
+
+    @Input()
+    isLoading = false;
+    extractParticipantName = (participantScoreDTO: ParticipantScoreDTO) => {
+        if (participantScoreDTO.userName) {
+            return `${participantScoreDTO.userName}`;
+        } else {
+            return `${participantScoreDTO.teamName}`;
+        }
```
Comment (derLalla) [inline src/main/webapp/app/shared/participant-scores/participant-scores-table/participant-scores-table.component.ts:19]:
```suggestion
        return participantScoreDTO.userName ? String(participantScoreDTO.userName) : String(participantScoreDTO.teamName);
```

**Review by derLalla — CHANGES_REQUESTED** [review]
this PR gets more and more stuffed 😅  It won't be easy to review all the 4000 lines of code. But I started with your latest commits and found some minor improvements :)

File: src/main/java/de/tum/in/www1/artemis/web/rest/dto/ParticipantScoreDTO.java [hunk 28]
```diff
@@ -0,0 +1,94 @@
...
+     */
+    public static ParticipantScoreDTO generateFromParticipantScore(ParticipantScore participantScore) {
+        String userName = null;
+        Long userId = null;
+        String teamName = null;
+        Long teamId = null;
+
+        if (participantScore.getClass().equals(StudentScore.class)) {
+            StudentScore studentScore = (StudentScore) participantScore;
+            userName = studentScore.getUser() != null && studentScore.getUser().getLogin() != null ? studentScore.getUser().getLogin() : null;
```
[the remark this answers is not in the document]
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/web/rest/dto/ParticipantScoreDTO.java:73]: I find the tertiary operator more understandable than relying on the return value of the conditional operators

File: src/main/java/de/tum/in/www1/artemis/web/rest/dto/ParticipantScoreDTO.java [hunk 29]
```diff
@@ -0,0 +1,94 @@
...
+            userId = studentScore.getUser() != null ? studentScore.getUser().getId() : null;
+        }
+        else {
+            TeamScore teamScore = (TeamScore) participantScore;
+            teamName = teamScore.getTeam() != null && teamScore.getTeam().getName() != null ? teamScore.getTeam().getName() : null;
```
[the remark this answers is not in the document]
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/web/rest/dto/ParticipantScoreDTO.java:78]: I find the tertiary operator more understandable than relying on the return value of the conditional operators

File: src/main/webapp/app/course/course-scores/course-scores.component.ts [hunk 30]
```diff
@@ -139,14 +145,53 @@ export class CourseScoresComponent implements OnInit, OnDestroy {
...
+                console.log(`No Course Score exists for regular calculation: ${JSON.stringify(regularCalculation)}`);
+            } else {
+                if (courseScoreDTO.pointsAchieved !== regularCalculation.overAllPoints) {
+                    console.log(
+                        `Different points in course scores dto. Regular Calculation: ${JSON.stringify(regularCalculation)}. Course Scores DTO : ${JSON.stringify(courseScoreDTO)}`,
+                    );
+                }
+                if (courseScoreDTO.scoreAchieved !== regularCalculation.overallScore) {
+                    console.log(
+                        `Different score in course scores dto. Regular Calculation: ${JSON.stringify(regularCalculation)}. Course Scores DTO : ${JSON.stringify(courseScoreDTO)}`,
```
[the remark this answers is not in the document]
Comment (stefanwaldhauser) [reply src/main/webapp/app/course/course-scores/course-scores.component.ts]: Done :) I throw a sentry error now as discussed with @krusche

**Comment by stefanwaldhauser** [comment]
> this PR gets more and more stuffed 😅 It won't be easy to review all the 4000 lines of code. But I started with your latest commits and found some minor improvements :)

The ResultListener is the most important part. You can ignore the rest in your review

File: src/test/javascript/spec/component/exam/exam-scores/exam-scores.component.spec.ts [hunk 31]
```diff
@@ -18,6 +18,8 @@ import { FaIconComponent } from '@fortawesome/angular-fontawesome';
 import { SortService } from 'app/shared/service/sort.service';
 import { JhiLanguageHelper } from 'app/core/language/language.helper';
 import { HelpIconComponent } from 'app/shared/components/help-icon.component';
+import { ParticipantScoresService, ScoresDTO } from 'app/shared/participant-scores/participant-scores.service';
+import * as _ from 'lodash';
```
Comment (iozbeyli) [inline src/test/javascript/spec/component/exam/exam-scores/exam-scores.component.spec.ts:22]: You are only using cloneDeep. Why are you importing the whole thing? 🤔

**Comment by stefanwaldhauser** [comment]
I updated the migration script as discussed with @krusche . 
With the new migration script there were no score / points differences in the course-scores calculation in PGDP, Patterns and other courses. 
image

File: src/test/javascript/spec/component/exam/exam-scores/exam-scores.component.spec.ts (hunk 31 above)
Comment (stefanwaldhauser) [reply src/test/javascript/spec/component/exam/exam-scores/exam-scores.component.spec.ts]: Changed to named import in 8df83725325b424a16ff384e5130eda5e17e2565

**Review by krusche — DISMISSED** [review]

**Review by krusche — APPROVED** [review]
